### ISSUE TITLE
 createrepo_c: non printable character in provides tag leads to invalid xml

### ISSUE THREAD

**Issue by M0ses** [rpm-software-management/createrepo_c#104]
As reported by @lnussel in

https://bugzilla.opensuse.org/show_bug.cgi?id=1110914

An rpm with a non printable character in it's provides tag leads createrepo_c to create invalid XML which makes other tools barf:

<rpm:entry name="ksym(default:^B)"/>

**Comment by lnussel** [comment]
what's the status here?

**Comment by M0ses** [comment]
ping!

**Comment by kontura** [comment]
I have been trying to solve this issue, but I got stuck.
I am not sure how should createrepo_c behave when this happens.

Original createrepo just ignored the characters and simply dropped them, this however leads to wrong repodata and DNF could have trouble with packages like this (it depends on where the character was in the original rpm, but dropping characters from file names, provides, requires... not good)

On the other hand I tried escaping these characters but this still leads to invalid xmls. Basically every tool I tried on such repodata (xmllint/xmlwf) complains. DNF ignores such a repo entirely, just like when the characters are no escaped. 
According to  [this document](https://www.w3.org/TR/xml/#charsets) only #x9, #xA and #xD (tab, line feed and carriage return) are allowed from non printable characters in the text data of xml.

The only option that I can think of is not including packages with such characters in the resulting repodata.
Or am I wrong somewhere? Is there some other way we could do this?
I would be grateful for any help.

**Comment by M0ses** [comment]
> I have been trying to solve this issue, but I got stuck.
> I am not sure how should createrepo_c behave when this happens.
> 
> Original createrepo just ignored the characters and simply dropped them, this however leads to wrong repodata and DNF could have trouble with packages like this (it depends on where the character was in the original rpm, but dropping characters from file names, provides, requires... not good)

I fully agree!

> 
> On the other hand I tried escaping these characters but this still leads to invalid xmls. Basically every tool I tried on such repodata (xmllint/xmlwf) complains. DNF ignores such a repo entirely, just like when the characters are no escaped.
> According to [this document](https://www.w3.org/TR/xml/#charsets) only #x9, #xA and #xD (tab, line feed and carriage return) are allowed from non printable characters in the text data of xml.
> 
> The only option that I can think of is not including packages with such characters in the resulting repodata.
> Or am I wrong somewhere? Is there some other way we could do this?

IMHO createrepo_c should exit with an error as early as possible to make this error visible to "users" aka package maintainers. Otherwise it might become really hard to find the root case.

Think of an released repo - why is my package not include - everthing else is inside, but not my package - how could I find out that my package includes binary char?

> I would be grateful for any help.

THX for taking care of this issue!

**Comment by kontura** [comment]
So the way it works now is if there is an error during handling of any specific package we print out the error. For example: 
`C_CREATEREPOLIB: Warning: Cannot read package...`
and continue because by the time we can do anything about it most of the work is already done (all the packages have been handled).

Even with such a broken rpm createrepo_c ends successfully, however there is an option `--error-exit-val` which if specified ensures that createrepo_c ends with exit value 2 if there were any problems during processing of rpms. 

It would make sense to me to keep this behavior consistent even in this case. We would print out some error msg like:
`C_CREATEREPOLIB: Critical: Cannot dump XML for abcde.rpm: Forbidden control chars found (ASCII value <32 except 9, 10 and 13).`
and continue, but if `--error-exit-val` specified we would exit with 2.

Does that seem reasonable?

**Comment by lukash** [comment]
Just my personal opinion, but if there were errors, I think the binary should return a value indicating an error.

Imagine running this automatically on some repository data. Someone may break the data and you'll never know, the repo will be generated just like nothing's wrong. The only way for this to be safe is to return an error, that should prod someone to fix it. The current behaviour seems unreliable to me and a source for trouble.

**Comment by M0ses** [comment]
@lukash : I fully agree

The problem occurred before (while generating the spec file) but it might become very hard to debug afterwards - tools like OBS (and maybe koji) will hide these problems from users and will raise questions afterwards (why is my package not included?) - which may be really hard to debug.

Please lets fail early!

**Comment by kontura** [comment]
Some of the reasoning behind this is described in https://github.com/rpm-software-management/createrepo_c/pull/63

So while it may not be ideal I think we should be consistent, it doesn't seem right to hard fail because of a wrong character but not in the case of completely empty.rpm file.

**Comment by lukash** [comment]
@kontura I can't find anything about an empty .rpm specifically in that PR, so don't know about that one... Seems to me they didn't want to fail building because of an error in a single package. So they agreed to introduce a special error status indicating there were errors even though the repodata were still built. But then, without further discussion, the PR author added a switch to enable this behaviour, instead of the to me more logical disable switch. So he didn't "fix" it by default, only with a switch.

I don't agree with that, I think it hides errors in the default mode and it shouldn't be so. Of course build tools need to be able to handle this, but I'd say quite surely the correct behaviour for build tools is still to return the error to indicate the status and allow them to take whatever action is correct in the context. I wouldn't be surprised if they were running with the `-error-exit-val` switch so that they can indicate the errors properly.

If the "empty .rpm" case is a special one that you always want to ignore, maybe it deserves special treatment, even though consistency is generally better, of course :slightly_smiling_face:.

**Comment by kontura** [comment]
@lukash I meant the empty rpm as an example there can be any number of errors while handling one specific rpm. I just wanted to point out that we should mostly handle them the same way including the non-printable character one.

About the `--error-exit-val` I think you are right, but I also think that if we changed that now it could break something. I would suggest separate issue/bugzilla for that as it is a bigger change?

So what I am trying to say is that we should behave as I described in https://github.com/rpm-software-management/createrepo_c/issues/104#issuecomment-499798714 and if it is later decided to change the default behavior it will also change this specific error handling because it will use the same mechanism.
